Sources/GAnalytics/GAnalytics.swift

```diff
@@ -8,41 +8,55 @@ import Logging
 /// An abstraction class for bringing Google Analytics (Firebase and Crashlytics) into the app from the Firebase package.
 /// To provide user-specific insights for logging app metrics and performance.
 public class GAnalytics {
+    private let app: AnalyticsApp.Type
+    private let analytics: AnalyticsLogger.Type
+    private let crashLogger: CrashLogger
+    private let preferenceStore: AnalyticsPreferenceStore
+    
     /// Additional parameters for the application
     public var additionalParameters = [String: Any]()
-    let analytics: AnalyticsLogger.Type
+    
+    init(app: AnalyticsApp.Type,
+         analytics: AnalyticsLogger.Type,
+         crashLogger: CrashLogger,
+         preferenceStore: AnalyticsPreferenceStore) {
+        self.app = app
+        self.analytics = analytics
+        self.crashLogger = crashLogger
+        self.preferenceStore = preferenceStore
+    }
+    
+    public convenience init() {
+        self.init(app: FirebaseApp.self,
+                  analytics: Analytics.self,
+                  crashLogger: Crashlytics.crashlytics(),
+                  preferenceStore: UserDefaultsPreferenceStore())
+    }
     
     /// Initialises the Firebase instance when launching the app.
     public func configure() {
-        FirebaseApp.configure()
-        bundleDataObserver()
-        fetchSettingBundleData()
+        app.configure()
+        subscribeToPreferenceStore()
+        updateAnalyticsPreference(preferenceStore.hasAcceptedAnalytics)
     }
     
-    /// Creates Notification Center Observer for user's analytics permissions for the app.
-    private func bundleDataObserver() {
-        NotificationCenter.default.addObserver(self, selector: #selector(fetchSettingBundleData), name: UserDefaults.didChangeNotification, object: nil)
+    private func subscribeToPreferenceStore() {
+        Task {
+            for await value in preferenceStore.stream() {
+                updateAnalyticsPreference(value)
```

Comment:
Trying to wrap my head around why we call this method twice as part of `configure()`... I think I might be getting it but just to clarify:
We're setting our analytics preference in Firebase based on what the user has set in UserDefaults when they first launch the app and then subscribe to any changes in that UserDefaults setting during the lifecycle of the app which will then update that preference in Firebase..?

Reply:
Yeah, will talk you through this in person, but the task runs constantly in the background asynchronously. As you say: when the stream outputs new values, that is to say the user changes their preferences, we update the preference in Firebase.